File: ObjectInit/ObjectInit/ObjectInit/Tire/Tire.m
//
//  Tire.m
//  ObjectInit
//
//  Created by 김기현 on 2021/01/20.
//

#import "Tire.h"


@implementation Tire

- (id) init {
    if (self = [super init]) {
        pressure = 34.0;
        treadDepth = 20.0;
    }
    
    return (self);
}

- (id) initWithPressure: (float) p treadDepth: (float) td {
    if (self = [super init]) {
        pressure = p;
        treadDepth = td;
    }
    
    return (self);
} // initWithPressure:treadDepth

// pressure를 값을 넘겨받아 설정하고 treadDepth는 20.0으로 초기화
- (id) initWithPressure: (float) p {
    if (self = [super init]) {
        pressure = p;
        treadDepth = 20.0;
    }
    
    return (self);
} // initWithPressure:

// pressure는 34.0으로 초기화하고 treadDepth는 값을 넘겨받아 설정
- (id) initWithTreadDepth: (float) td {
    if (self = [super init]) {
        pressure = 34.0;
        treadDepth = td;
    }
    
    return (self);
} // initWithTreadDepth:

- (void) setPressure: (float) p {
    pressure = p;
}

- (float) pressure {
    return (pressure);
}

- (void) setTreadDepth: (float) td {
    treadDepth = td;
}

- (float) treadDepth {
    return (treadDepth);
}

- (NSString *) description {
    NSString *desc;
    desc = [NSString stringWithFormat: @"Tire: Pressure: %.1f TreadDepth: %.1f", pressure, treadDepth];
    
    return (desc);
}

@end // Tire

File: CarProtocol Swift/CarProtocolSwift/CarProtocolSwift/Tire/Tire.swift
//
//  Tire.swift
//  CarProtocolSwift
//
//  Created by 김기현 on 2021/02/17.
//

import Foundation

class Tire: NSCopying {
    
    var pressure: Float
    var treadDepth: Float
    
    convenience init() {
        self.init(pressure: 34, treadDepth: 20)
    }
    
    required init(pressure p: Float, treadDepth td: Float) {
        self.pressure = p
        self.treadDepth = td
    }
    
    convenience init(p: Float) {
        self.init(pressure: p, treadDepth: 20.0)
    }
    
    convenience init(td: Float) {
        self.init(pressure: 42.0, treadDepth: td)
    }
    
    func copy(with zone: NSZone? = nil) -> Any {
        let tireCopy: Tire?
        tireCopy = type(of: self).init(pressure: pressure, treadDepth: treadDepth)
        
        guard let result = tireCopy else { fatalError("copy is fail") }
        return result
    }
    
    func setPressure(p: Float) {
        pressure = p
    }
    
    func getPressure() -> Float {
        return pressure
    }
    
    func setTreadDepth(td: Float) {
        treadDepth = td
    }
    
    func getTreadDepth() -> Float {
        return treadDepth
    }
    
    func description() -> String {
        let desc: String = String(format: "Tire Pressure: %.1f TreadDepth: %.1f", pressure, treadDepth)
        
        return desc
    }
}
